plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -74,92 +88,252 @@ The current value of the underlying asset is determined by an oracle. See note
+    => Role
+    -> m ()
+futureContract r =
+    let adjustMarginEP = liftBase (endpoint "adjust margin") >>= adjustMarginC r
+        settleEarlyEP  = liftBase (endpoint "settle early") >>= settleEarlyC
+        settleEP       = liftBase (endpoint "settle") >>= settleC
+        go = do
+                alive <- selectEither updateState (adjustMarginEP <|> settleEarlyEP <|> settleEP)
+                case alive of
+                    Left Dead -> pure ()
+                    _         -> go
+    in go
+
+-- | Wait for a change to the contract state on-chain, update the
+--   state, then return 'Liveness' value indicating whether the contract
+--   is still going.
+updateState
+    :: ( MonadFuture r m, Member WatchAddress r )
+    => m Liveness
+updateState = do
+    addr <- contractAddress
+    tx <- liftBase (nextTransactionAt addr)
+    -- We have a transaction 'tx' that modifies the UTXO set at the contract
+    -- address.
+    -- We now need to update the 'futureData' field of the state with the new
+    -- 'FutureData'. We could get the new 'FutureData' by unlifting the data
+    -- script from script output of 'tx', BUT we can't unlift things yet. But
+    -- we can infer what the new state is from the following facts.
+    --
+    -- 1. The only two state-changing actions are 'AdjustMargin' and 'Settle'.
+    -- 2. If the action performed was 'Settle' then 'tx' does not have any
+    --    outputs at the script address because the instance is done.
+    -- 3. If the action was 'AdjustMargin', then either the
+    --    'futureDataMarginShort' or the 'futureDataMarginLong' field has
+    --    changed. The delta is exactly the delta of Ada that was locked by the
+    --    contract previously vs the amount that is locked by 'tx'.
+
+    oldState <- gets futureData
+    lockedPrior <- gets knownOutputs >>= totalValueLocked
+    lockedNow   <- totalValueLocked (AM.fromTxOutputs tx)
+    let delta = Ada.fromValue lockedNow - Ada.fromValue lockedPrior
+        newDS = listToMaybe $ mapMaybe LTx.txOutData $ view LTx.outputs tx
+        fd1   = oldState { futureDataMarginLong = futureDataMarginLong oldState + delta }
+        fd2   = oldState { futureDataMarginShort = futureDataMarginShort oldState + delta }
+    case newDS of
+        Just ds'
+            | ds' == DataScript (Ledger.lifted fd1) -> do
+                modify $ \s -> s { futureData = fd1, knownOutputs = AM.updateAddresses tx (knownOutputs s) }
+                pure Alive
+            | ds' == DataScript (Ledger.lifted fd2) -> do
+                modify $ \s -> s { futureData = fd2, knownOutputs = AM.updateAddresses tx (knownOutputs s) }
+                pure Alive
+            | ds' == DataScript (Ledger.lifted oldState) -> do
+                -- Nothing changed (this happens when the contract is first
+                -- initialised)
+                modify $ \s -> s { knownOutputs = AM.fromTxOutputs tx }
+                pure Alive
+        _ -> pure Dead
```

Comment:
If we modeled this as a state machine (is there a reason we aren't?), we could provide a standard function to monitor the state machine and replicate its action locally, thus allowing you to see what the current state is.

Reply:
Good point, we can probably share more code between on-chain and off-chain code that way.

The reason why this isn't a state machine is that the contract was implemented before we had state machines. (That's why it doesn't check the next data script which is actually a serious flaw in the contract :o ) I'll update it.